applications/SeismicApplication/custom_constitutive/uniaxial_kent_park.cpp

```diff
@@ -0,0 +1,252 @@
+{
+    KRATOS_TRY
+    double strain0  = -1.0 * rValues.GetMaterialProperties()[CONCRETE_YIELD_STRAIN];
+    double fc       = -1.0 * rValues.GetMaterialProperties()[CONCRETE_YIELD_STRENGTH];
+    double strain_u = -1.0 * rValues.GetMaterialProperties()[CONCRETE_CRUSHING_STRAIN];
+    // pre yielding
+    if (rStrainVector[0] > strain0) {
+        double eta = rStrainVector[0] / strain0;
+        rStressVector[0] = fc * (2.0*eta - eta*eta);
+        double E0 = 2.0 * fc / strain0;
+        mTangentModulus = E0 * (1.0 - eta);
+    // softening
+    } else if (rStrainVector[0] >= strain_u) {
+        mTangentModulus = 0.8*fc / (strain0-strain_u);
+        rStressVector[0] = fc + mTangentModulus*(rStrainVector[0]-strain0);
+    // past ultimate strain
+    } else {
+        rStressVector[0] = 0.2*fc;
+        mTangentModulus = 0.0;
+    }
+    KRATOS_CATCH("")
+}
+
+void UniaxialKentParkMaterialLaw::Unload(
+    Parameters& rValues, Vector& rStrainVector, Vector& rStressVector)
+{
+    KRATOS_TRY
+    double strain0  = -1.0 * rValues.GetMaterialProperties()[CONCRETE_YIELD_STRAIN];
+    double fc       = -1.0 * rValues.GetMaterialProperties()[CONCRETE_YIELD_STRENGTH];
+    double strain_u = -1.0 * rValues.GetMaterialProperties()[CONCRETE_CRUSHING_STRAIN];
+    double eps_tmp = mStrainR;
+    if (eps_tmp < strain_u) {
+        eps_tmp = strain_u;
+    }
+    double eta = eps_tmp / strain0;
+    double ratio = 0.707 * (eta-2.0) + 0.834;
+    if (eta < 2.0) {
+        ratio = 0.145*eta*eta + 0.13*eta;
+    }
+    mStrainP = ratio * strain0;
+
+    double E0 = 2.0 * fc / strain0;
+    double tmp1 = mStrainR - mStrainP;
+    double tmp2 = rStressVector[0] / E0;
+    if (tmp1 > -std::numeric_limits<double>::epsilon()) {
+        mUnloadSlope = E0;
+    } else if (tmp1 <= tmp2) {
+        mStrainP = mStrainR - tmp1;
+        mUnloadSlope = rStressVector[0] / tmp1;
+    } else {
+        mStrainP = mStrainR - tmp2;
+        mUnloadSlope = E0;
+    }
+    KRATOS_CATCH("")
+}
+
+void UniaxialKentParkMaterialLaw::CalculateConstitutiveMatrixPK2(
+    Parameters& rValues, Matrix& rConstitutiveMatrix)
+{
+    KRATOS_TRY
+    rConstitutiveMatrix(0,0) = mTangentModulus;
+    KRATOS_CATCH("")
+}
+
+std::string UniaxialKentParkMaterialLaw::Info() const {
+    std::stringstream buffer;
+    buffer << "UniaxialKentParkMaterialLaw";
+    return buffer.str();
+}
+
+void UniaxialKentParkMaterialLaw::PrintInfo(std::ostream& rOStream) const
+{
+    rOStream << "UniaxialKentParkMaterialLaw";
+}
+
+void UniaxialKentParkMaterialLaw::PrintData(std::ostream& rOStream) const
+{
+}
+
+void UniaxialKentParkMaterialLaw::save(Serializer& rSerializer) const
+{
+    KRATOS_SERIALIZE_SAVE_BASE_CLASS(rSerializer, ConstitutiveLaw);
+    rSerializer.save("mStrainR", mStrainR);
+    rSerializer.save("mStrainP", mStrainP);
+    rSerializer.save("mUnloadSlope", mUnloadSlope);
+    rSerializer.save("mTangentModulus", mTangentModulus);
```

Comment:
Do you really need to save everything?, particularly the tangent modulus

Reply:
The tangent modulus, in particular, I am saving in memory because the computation of the material response involves many steps so that we do not have to do all the computations when we just need the tangent modulus. I am looking into reducing the number of saved variables ..

Reply:
I trust you, just in case

Reply:
I reduced many of the mConverged* variables, and I believe the results are still the same. :crossed_fingers:

Reply:
I recommend to create a test before change anything and change the things later, so you know you haven't broke anything

Reply:
Sorry for the delay. I fixed the test case and added it to CI, but the test case fails and the results are slightly different. So, I think keeping the extra variables is needed for now :neutral_face: